### PR TITLE
Apply result of calculations with copy on change properties

### PR DESCRIPTION
**Pull request by aclex**
## Problem

'Copy on change' properties propagated to SubShapeBinder can be changed there and then updated in the base body (support). But they also might be used as a variable in calculation of some other 'copy on change' properties of the base body which are propagated to the same SubShapeBinder as well. It is expected to see all the 'copy on change' properties in the SubShapeBinder recalculated based on the modified values of propagated properties and using the expressions defined in the base (support) body.

## Example

Here's an example project (seems I can't attach "*.FCStd" file, will describe it here thoroughly).

It contains **Body1** with three user-defined properties: ***Base_Depth*** (set to 200 mm), ***Base_Width*** (with expression `= Base_Depth * 3`) and ***Base_Thickness*** (with expression `= Base_Depth / 10`). All these properties marked with 'CopyOnChange' status. ***Base_Depth*** and ***Base_Width*** used in a sketch to define the dimensions of a rectangle (using expressions with `href()` to reference upstream properties). There's a pad using this sketch to extrude it for the length of ***Base_Thickness***.

The project also has SubShapeBinder **Binder** created from **Body1** (it has **Body1** as its `Support` link) with 'Bind Copy On Change' property initially set to 'Enabled'. Thus it has all the three user-defined properties mentioned above propagated.

Now I change the value of ***Base_Depth*** property in the **Binder** to 300 mm instead of 200 mm ('Bind Copy On Change' property is changed to 'Mutated' here). I see the part recalculated and repainted properly compared to the original **Body1**.
shot_20240705_013907
But I also expect ***Base_Width*** and ***Base_Thickness*** of **Binder** to be updated properly (to 900 mm and 30 mm, correspondingly), which is not the case. They keep containing the original values initially taken from **Body1**.
shot_20240705_014049

To futher experiment with it I create **Body2** out of **Binder** (as a 'Base Feature'), add another sketch to it to make a pocket. I use ***Base_Width*** of the base feature to define the length of the pocket expecting it to cover all the width of the part. But that's not the case, as ***Base_Width*** is 600 mm here as well and the pocket is done only in the middle.
изображение

## Fix

For the fix I've added another loop copying all the copy on change properties from temporary copy of document back to the original just after the recalculation. Direct copy looks safe to me here, as all these properties were copied to the temporary right before its recalculation.

I find this little fix very important in my use case as I try to create some prototype part with a suite of top-level input parameters (some key dimensions) and output values (e.g. outer dimensions of complex part) to be used as base for several concrete parts. Now this scenario is hardly possible as the most-derived part is just partially affected by the parameter changes and values may or may not be correct.

### CODE REVIEW

**Comment by chennes** [comment]
Thanks for the PR -- sorry for the delay getting tat CI run started. It's going now...

**Comment by aclex** [comment]
No problem, thank you for taking a look at it!

**Comment by chennes** [comment]
Can you rename your example FCStd file to *.zip and try to upload it? I think that will work. If not I'll follow your instructions, I'm just trying to be a bit lazy here...

**Comment by aclex** [comment]
Yes, sure, here it is: [copy_on_change_properties_behaviour.zip](https://github.com/user-attachments/files/16261702/copy_on_change_properties_behaviour.zip)

File: src/Mod/PartDesign/App/ShapeBinder.cpp [hunk 1]
```diff
@@ -621,6 +621,20 @@ void SubShapeBinder::update(SubShapeBinder::UpdateOption options) {
...
+                            auto p = copied->getPropertyByName(prop->getName());
+                            if (p && p->getContainer() == copied
+                                && p->getTypeId() == prop->getTypeId()
+                                && !p->isSame(*prop))
+                            {
+                                std::unique_ptr<App::Property> pcopy(p->Copy());
+                                prop->Paste(*pcopy);
+                            }
+                        }
+                    }
```
Comment (chennes) [inline src/Mod/PartDesign/App/ShapeBinder.cpp:637]: This code block is basically identical to the block above -- at a minimum it should be refactored so we don't repeat ourselves.

**Review by chennes — CHANGES_REQUESTED** [review]
Overall this seems to only kick the can down the road by doing a single pass of additional updating. What happens if there is a longer dependency chain?

**Comment by aclex** [comment]
Yes, currently it only sort of 'completes' one pass of changes. I thought of it and considered it OK just because cyclic references are prohibited in calculations elsewhere anyway, so it's quite in style with the whole approach, with same restrictions (one pass only). The general approach here seems to be kind of gradient descent (i.e. apply cycle till there's no more changes or their amount are less than ε), but I don't think, it's the right place to adopt something like this for the first time :)

File: src/Mod/PartDesign/App/ShapeBinder.cpp (hunk 1 above)
Comment (aclex) [reply src/Mod/PartDesign/App/ShapeBinder.cpp:637]: Could you please suggest the right solution for it to be in style with the rest of the code? Is it enough to get it to lambda and call it from a couple of places? The context is the same in both cases.
As for the duplication, I just didn't want to touch the existing code, thought the patch is 'more minimal' this way :)

**Comment by chennes** [comment]
@wwmayer code duplication aside what are your thoughts on this approach?

**Comment by wwmayer** [comment]
Using a lambda function would be a way to go.

**Comment by aclex** [comment]
Nice, thank you! Will fix it this evening.

**Comment by pieterhijma** [comment]
First of all, thanks @aclex for the PR that shows an interesting problem with copy-on-change values. However, looking at this PR I recognize some issues with this solution. 

Firstly, I see some issues with the semantics. `Body1` has only `Depth` that drives the design, `Thickness` and `Width` depend on it. I understand why to make `Depth` copy-on-change, but I don't see why `Thickness` and `Width` should be copy-on-change.

I might understand the need to let `Thickness` and `Width` be some expression as a default value unless you decide to "override" that with a custom value, but this semantics is not supported by the current copy-on-change logic, at least not in this way.

Perhaps it is good to try to understand what happens:

As soon as one makes a subshape binder, it links to the original "support" and the subshape binder inherits the copy on change properties. As soon as you change one of those properties, a copy of the support is created in a temporary hidden document and the copy-on-change properties are copied into this object, which includes the original expression (by means of a separate property Expression Engine). However, the subshape binder still shows the original value because that wasn't changed. Because subshapebinders' goal is to keep track of properties that are marked copy-on-change, what happens somehow makes sense to me. Additionally, in support of this (or these shortcomings) the binder doesn't show the expression that drives this property.

So, the binder keeps track of the copy-on-change properties. The user can change copy on change properties with expressions in them all they want, but they will be overridden by the expression that governs the value of the property. That is why the shape has the "right" dimensions in the example.

If anything, I would prefer the reverse to happen for copy-on-change properties that have expressions: as soon as such a property is changed, the expression is removed from the shadow copy and it takes on the value that the user gave.

However, I prefer the current semantics with a rule-of-thumb: properties with expressions should not be marked to be copy-on-change because of this very behavior, let's call it undefined behavior.

So, I would propose to document somewhere that copy-on-change should not be driven by expressions.

Secondly, this PR doesn't allow the user to "override" a copy-on-change property that contains an expression. Adjusting `Base_Thickness` to some other value doesn't result in any change (because the properties are governed by an expression) and the value jumps back to the value computed by the expression.

So, all in all, I would suggest to not merge this PR and close it. For a broader discussion, I would suggest to create an issue. 

Practically, I simply suggest to remove the copy-on-change values on the two properties with expressions and all will work well, I believe, except for some references that should be more general by having them refer to either `Body1` (or `<<BaseFeature>>.BaseFeature.Support[0][0]` for example) and then expressed in terms of the driving property `Base_Depth`. The drawback is that you need to repeat the relations but this is a good reason why FreeCAD has Spreadsheets or VarSets to define fundamental relations that can be repeated throughout the design.

For the argument that a user may want to have expressions as default values that could be overridden, I would say that subshape binders are not the right solution. If a design relies on relations like were defined (`base_length * 3, base_length / 10`) and the goal is to be able to parameterize on that by removing one or more of those relations, I would say the design becomes a fundamentally different one that should have its own sketch as base. Or, I would recommend to define those relations more flexibly with VarSets or Spreadsheets as mentioned above.

An additional benefit of using VarSets or Spreadsheets is that you circumvent the circular dependency problems.

**Comment by aclex** [comment]
@pieterhijma, many thanks for such an in-depth comment on my request! And also thank you very much for such a friendly discussion, much appreciated. I'd like to reply on some points you've mentioned, if you please.

At first, yes, just to clarify my initial intention a litte more, I want to have a 'prototype' part, which:
  a) contains (or even encapsulates) some complex calculations and dependencies based on several 'input' parameters (properties) and puts the results in several 'output' parameters, and
  b) is subject to further fixes and changes, so that all the bound parts are updated automatically

So yes, I completely agree that while these 'input' parameters (`depth` in my example) are truly copy-on-change, 'output' parameters (`width` and `thickness` here) have actually nothing to do with copy-on-change, except (correct me if I'm wrong) making them copy-on-change is currently the only way to have them propagated to ShapeBinder and SubshapeBinder created on the body. In fact, I think, these properties almost always should be `Output` and `ReadOnly`, and never `CopyOnChange`, can't really imagine any scenario for them even to be writable. So another simple solution to this, from my point of view, could be a rule, that e.g. properties which are both `Output` and `ReadOnly` are propagated to binders as well as `CopyOnChange` ones. Looks like that would just fix the problem completely. And also it seems to be easy to implement even during this request, just changing the condition on properties to be copied back from base copy to derived document.

The reason I've still decided to come up with pull request rather than discuss it on the forum firstly is the inconsistency I've described in the header, when the complex body becomes partially calculated using initial values and on another part using the modified values. If I understand you correctly, this is exactly what you propose to call undefined behavior and not recommended use, which I tend to agree. Copy-on-change parameter with expression is indeed quite confusing and controversial scenario, but yes, again, the copy-on-change attribute here is just to make the property propagated to the binders, not for its proper use.

As for other possible workarounds for this problem you've mentioned, maybe I miss something, especially for version 1.0, as I've only tried spreadsheets to make some similar thing, but I think it's also quite a restricted approach, too, as I seem to can't 'bind' spreadsheet to any body or binder I wish to make it get arbitrary parameters for expressions in cells from the bound object and make its calculations based on them. Configuration tables seems to address quite similar problem, but, as I can see, they are still tables by their nature, so they map enumeration values to real values, rather than real values to real values.

So, to sum up: can we discuss the approach of the properties with some other selected attributes (I propose Output and Read-only) are propagated to ShapeBinder and SubshapeBinder, either here or in separate issue? It looks to me as the optimal solution for the problem.

**Comment by aclex** [comment]
I've just pushed some correction to show the approach I propose more clearly. I didn't change the copy-on-change attribute indicating the property is propagated to SubShapeBinder as it would be quite a big change. But I added a condition for a property to be copied from Support body back to binder after recalculation: it should be both 'Output' and 'ReadOnly'. As far, as I can see, this additional requirement addresses all the issues which @pieterhijma pointed out. Hopefully it's now consistent enough.

I would also want to ask to still consider this change for the next major release (1.0, as I can see), if it's possible, as it seems quite important, yet minor, change to me.

Again, it's not about default pre-calculated values or some other corner cases with value overriding expression etc. This is actually what makes the idea of SubshapeBinder more complete and allow to create a library of reusable base bodies. Now the user can get a number of similar parts using SubshapeBinder each slightly different in customization points defined in the base body. The user can even create another body on top of SubshapeBinder, with customizations. But currently it's not possible to use any parameter from the base body in further calculations of the derived body, if any value customization point is changed. In my opinion, it kind of hurts the whole this nice concept.

**Comment by pieterhijma** [comment]
Hi @aclex, Thanks, I understand better now. I'm looking into the PR.

> I would also want to ask to still consider this change for the next major release (1.0, as I can see), if it's possible, as it seems quite important, yet minor, change to me. 

@maxwxyz, @chennes, perhaps take another look into whether it is 1.0 or not?

**Comment by pieterhijma** [comment]
Looked at it and the change makes sense to me. Let me try to summarize what the PR does:

With this PR we can transfer 2 types of copy-on-change parameters to subshapebinder or links:
- "true" copy-on-change parameters that users can use to change the geometry (was already available before this PR).
- output/read-only copy-on-change parameter that users can base derivative geometry on.

This allows users to create a base geometry that can be extended by means of subshapebinders. These subshapebinders can be used as a base feature in other geometry. Since the subshapebinders now also have the output/read-only copy-on-change parameters, they can be used to build derivative geometry on. The output/read-only copy-on-change parameters may (but don't have to) be depend on the "true" copy-on-change parameters, therefore automatically updating the geometry.

I think this is a good way to improve modular and parametric design, so, I'm in favor of merging.

**Comment by aclex** [comment]
@pieterhijma, many thanks for revisiting the PR and your positive response on it! Your conclusions are exactly what I mean.

**Comment by maxwxyz** [comment]
@aclex any news on this PR?

**Comment by aclex** [comment]
@maxwxyz It's up-to-date version of the patch here, with all the fixes addressing the issues from previous review and @pieterhijma's suggestions. As I understand, now it has no release target and is still waiting for a review again.

I also continue testing it on my machine with a version compiled on top of 0.21.2 previously and 1.0rc2 now. I've noticed some problem with the warning messages like:
```
06:35:12  6.18204 <App> Document.cpp(2801): Recursive calling of recompute for document _tmp_binder
```
being printed when I work with the projects where binders and references in expressions are used intensively in nested parts (including 'href' links), with this new copy-on-change functionality applied in a number of places. I don't know, if it has something with it and plan to debug it more thoroughly while the request is silent when I have some time for it (in coming weeks, hopefully).

**Comment by chennes** [comment]
Sorry for the delay -- I'll give this a review this week.

**Comment by PaddleStroke** [comment]
@chennes any update on the review

**Comment by maxwxyz** [comment]
@chennes ping for review

**Comment by maxwxyz** [comment]
waiting for review

**Comment by maxwxyz** [comment]
@chennes ping for review

**Review by chennes — APPROVED** [review]
Thanks for your patience (and thanks to @pieterhijma for his thorough analysis). I think this is good to go now, pending the rebase and CI.

**Comment by aclex** [comment]
Many thanks for the merge and great discussion!
